scripts/backout_data_collection.py

```diff
@@ -0,0 +1,198 @@
+        return json.load(file)
+
+
+def preprocess_commits_and_bugs():
+    cache_path_commit = "dataset/cache_commit_dict.json"
+    cache_path_bug_to_commit = "dataset/cache_bug_to_commit_dict.json"
+    cache_path_bug = "dataset/cache_bug_dict.json"
+
+    if (
+        os.path.exists(cache_path_commit)
+        and os.path.exists(cache_path_bug_to_commit)
+        and os.path.exists(cache_path_bug)
+    ):
+        logger.info("Loading cached data...")
+        commit_dict = load_dict_from_file(cache_path_commit)
+        bug_to_commit_dict = load_dict_from_file(cache_path_bug_to_commit)
+        bug_dict = load_dict_from_file(cache_path_bug)
+
+    else:
+        logger.info("Preprocessing commits and bugs...")
+        commit_dict = {}
+        bug_to_commit_dict = {}
+
+        # store commits with their hashes and bug IDs as keys
+        for commit in tqdm(
+            repository.get_commits(
+                include_no_bug=True, include_backouts=True, include_ignored=True
+            ),
+            desc="Processing commits",
+        ):
+            commit_dict[commit["node"]] = {
+                "node": commit["node"],
+                "bug_id": commit["bug_id"],
+                "desc": commit["desc"],
+                "pushdate": commit["pushdate"],
+                "backedoutby": commit["backedoutby"],
+                "backsout": commit["backsout"],
+            }
+
+            if commit_dict[commit["node"]]["bug_id"] not in bug_to_commit_dict:
+                bug_to_commit_dict[commit["bug_id"]] = [commit_dict[commit["node"]]]
+            else:
+                bug_to_commit_dict[commit["bug_id"]].append(commit_dict[commit["node"]])
+
+        logger.info("Preprocessing bugs")
+        bug_dict = {}
+
+        num_lines = sum(1 for line in open(bugzilla.BUGS_DB, "r"))
+
+        # store bugs with their bug IDs as keys
+        with open(bugzilla.BUGS_DB, "r") as f:
+            for line in tqdm(f, total=num_lines, desc="Processing bugs"):
+                bug = json.loads(line)
+                bug_dict[bug.get("id")] = bug["resolution"]
+
+        save_dict_to_file(commit_dict, cache_path_commit)
+        save_dict_to_file(bug_to_commit_dict, cache_path_bug_to_commit)
+        save_dict_to_file(bug_dict, cache_path_bug)
+
+    return commit_dict, bug_to_commit_dict, bug_dict
+
+
+def filter_commits(
+    directory_path: str,
+    destination_filepath: str,
+    count_limit: int,
+    bug_dict: dict,
+    commit_dict: dict,
+    bug_to_commit_dict: dict,
+) -> None:
+    filtered_list = []
+    counter = 0
+
+    pbar = tqdm(total=count_limit, desc="Filtering commits")
+
+    for commit in repository.get_commits(
+        include_no_bug=False, include_backouts=True, include_ignored=False
+    ):
+        # add commit if it was backed out and the bug is fixed
+        if (
+            bug_dict.get(str(commit["bug_id"]))
+            and commit["backedoutby"]
+            and bug_dict[str(commit["bug_id"])] == "FIXED"
```

Comment:
storing the bug in a variable here will prevent retrieving it from the dict multiple times.